# Introdução ao PANDAS (continuação)
## IMDB dataset

In [1]:
import pandas as pd

df_IMDB = pd.read_csv("IMDB-Movie-Data.csv")

df_IMDB

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                 NaN       45.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
998                 93     5.6    4881                 NaN       22.0  
999                 87     5.3   12435               19.64       11.0  

[1000 rows x 12 columns]

### Revisão das funções já conhecidas:
* .head()
* .tail()

In [2]:
df_IMDB.head(2)

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   

                                         Description      Director  \
0  A group of intergalactic criminals are forced ...    James Gunn   
1  Following clues to the origin of mankind, a te...  Ridley Scott   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0

In [3]:
df_IMDB.tail(2)

Rank         Title                  Genre  \
998   999  Search Party       Adventure,Comedy   
999  1000    Nine Lives  Comedy,Family,Fantasy   

                                           Description          Director  \
998  A pair of friends embark on a mission to reuni...    Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...  Barry Sonnenfeld   

                                                Actors  Year  \
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
998                 93     5.6   4881                 NaN       22.0  
999                 87     5.3  12435               19.64       11.0

### Parâmetros do nosso DataFrame

* **index_col="title"** - parâmetro index
<p>Este serve para podermos ordenar as observações através de determinada coluna (neste caso a coluna "title").</p>
<p>Se não colocarmos este parâmetro, o pandas por defeito cria uma "coluna" inicial com os índices das observações.</p>

In [4]:
df_IMDB_index = pd.read_csv("IMDB-Movie-Data.csv",index_col="Title")

df_IMDB_index.head(5)

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

*Curiosamente o nosso dataframe tem uma coluna (rank) que ordena as nossas observações por um certo "índice" (1 a 1000).*

---

### Informações gerais do meu Dataframe
* **.info()** - função que nos dá muita informação; type e shape da dataframe + índices, nulos e datatype das colunas

In [5]:
df_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


*Podemos observar que nas colunas "Revenue" e "Metascore" temos várias observações sem informação (onull) para estas colunas.*

---

### Confirmar se existem dados Duplicados
* **.drop_duplicates()** - função que identifica e elimina as observações duplicadas (informações com dados iguais e gerados no mesmo momento).
<p>Parâmetros:</p>
<ul><li><strong>keep='first'</strong> - por defeito assume 'first', que mantém a informação da primeira observação duplicada e elimina os restantes.</li>
<li><strong>keep='last'</strong> - quando queremos manter a informação da última observação duplicada e elimina as restantes.</li>
<li><strong>inplace=False</strong> - por defeito assume False, não altera a própria dataframe, criar uma nova dataframe sem os duplicados;</li>
<li><strong>inplace=True</strong> - quando queremos fazer esta alteração no dataframe, os duplicados são eliminados no próprio dataframe original.</li></ul>

In [6]:
#inplace=False
df2_IMDB = df_IMDB.drop_duplicates()

df2_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


*Podemos observar que não houve alterações relativamente ao df_IMDB. Isto quer dizer que não temos nenhuma observação duplicada.*

In [7]:
#inplace=True
df_IMDB.drop_duplicates(inplace=True)

df_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


#### Concatenação
* **.concat()** - função que permite a concatenação de duas dataframes, neste caso iguais.

Desta forma estamos a duplicar as nossas observações e assim conseguimos identificar melhor o que a função **.drop_duplicates()** realiza.

In [8]:
df2_IMBD = pd.concat([df_IMDB, df2_IMDB])
df2_IMBD.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                2000 non-null   int64  
 1   Title               2000 non-null   object 
 2   Genre               2000 non-null   object 
 3   Description         2000 non-null   object 
 4   Director            2000 non-null   object 
 5   Actors              2000 non-null   object 
 6   Year                2000 non-null   int64  
 7   Runtime (Minutes)   2000 non-null   int64  
 8   Rating              2000 non-null   float64
 9   Votes               2000 non-null   int64  
 10  Revenue (Millions)  1744 non-null   float64
 11  Metascore           1872 non-null   float64
dtypes: float64(3), int64(4), object(5)
memory usage: 203.1+ KB


*Podemos observar que temos 2000 observações, o dobro das iniciais. O df2_IMDB está agora com 1000 observações duplicadas. Vamos confirmar usando a função .drop_duplicates()*

In [9]:
#Utilizamos novamente a função .drop_duplicates() para voltarmos a uma df sem duplicados
df2_IMDB.drop_duplicates(inplace=True)

df2_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


*Podemos observar que voltamos a ter as 1000 observações, sem qualquer duplicação.*

---

### Informações sobre as Colunas do nosso Dataframe
* **.columns** - atributo que podemos consultar no nosso dataframe para observarmos todos os nomes das colunas deste.

In [10]:
df_IMDB.columns

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

* **.rename()** - função que podemos utilizar para renomear o nome de uma ou mais colunas do nosso dataframe.

In [11]:
#Alterar o nome de 2 colunas
df_IMDB.rename(columns={
    'Runtime (Minutes)': 'Runtime',
    'Revenue (Millions)': 'Revenue' 
}, inplace=True)
df_IMDB.columns

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime', 'Rating', 'Votes', 'Revenue', 'Metascore'],
      dtype='object')

* **Maiúsculas vs Minúsculas** - podemos usar Python para converter os nomes das colunas para:
<ul><li>maiúsculas - col.upper()</li>
<li>minúsculas - col.lower()</li></ul>

In [12]:
#Podemos usar python para vermos o nome das nossas colunas
for col in df_IMDB.columns:
    print(col)

Rank
Title
Genre
Description
Director
Actors
Year
Runtime
Rating
Votes
Revenue
Metascore


In [13]:
#Podemos usar python para colocarmos os nomes das colunas em minúsculas
for col in df_IMDB.columns:
    print(col.lower())

rank
title
genre
description
director
actors
year
runtime
rating
votes
revenue
metascore


In [14]:
#Podemos usar python para colocarmos os nomes das colunas em maiúsculas
for col in df_IMDB.columns:
    print(col.upper())

RANK
TITLE
GENRE
DESCRIPTION
DIRECTOR
ACTORS
YEAR
RUNTIME
RATING
VOTES
REVENUE
METASCORE


* **Alterar todos os nomes das colunas** - podemos usar Python para converter os nomes das colunas para maiúsculas e minúsculas, utilizando listas novas e ciclos for para alterarmos como quisermos os nossos nomes das colunas.

In [15]:
#Criamos uma nova lista vazia que vamos povoar com os nomes das colunas em minísculas, através da função .append()
novos_nomes = []
for col in df_IMDB.columns:
    novos_nomes.append(col.lower())
print(novos_nomes)

#Atribuimos à lista de nomes de colunas da nossa df, a lista de nomes criada com minúsculas
df_IMDB.columns = novos_nomes
df_IMDB.columns

['rank', 'title', 'genre', 'description', 'director', 'actors', 'year', 'runtime', 'rating', 'votes', 'revenue', 'metascore']


Index(['rank', 'title', 'genre', 'description', 'director', 'actors', 'year',
       'runtime', 'rating', 'votes', 'revenue', 'metascore'],
      dtype='object')

In [16]:
#Confirmamos que a nossa df tem os nomes alterados para minúsculas
df_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   rank         1000 non-null   int64  
 1   title        1000 non-null   object 
 2   genre        1000 non-null   object 
 3   description  1000 non-null   object 
 4   director     1000 non-null   object 
 5   actors       1000 non-null   object 
 6   year         1000 non-null   int64  
 7   runtime      1000 non-null   int64  
 8   rating       1000 non-null   float64
 9   votes        1000 non-null   int64  
 10  revenue      872 non-null    float64
 11  metascore    936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [17]:
#Outra opção com menos código, neste caso nomes colunas em maiúsculas:
df_IMDB.columns = [col.upper() for col in df_IMDB.columns]
df_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   RANK         1000 non-null   int64  
 1   TITLE        1000 non-null   object 
 2   GENRE        1000 non-null   object 
 3   DESCRIPTION  1000 non-null   object 
 4   DIRECTOR     1000 non-null   object 
 5   ACTORS       1000 non-null   object 
 6   YEAR         1000 non-null   int64  
 7   RUNTIME      1000 non-null   int64  
 8   RATING       1000 non-null   float64
 9   VOTES        1000 non-null   int64  
 10  REVENUE      872 non-null    float64
 11  METASCORE    936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


---

### Dados nulos ou vazios
* **.isnull()** - cria um novo dataframe onde realiza uma ação booleana sobre todos os campos de dados do nosso dataframe original, determinando se esses campos tem dados ou não.

O que fazer com estes dados?

Temos várias estratégias que devemos ponderar com o objetivo de purificar os meus dados e reduzir o meu erro de análise. Vamos produzir sempre um erro, mas a escolha da melhor estratégia tem como objetivo produzir um erro menor possível.

A escolha de que estratégia escolher depende da minha análise ao dataframe, de como recolhi os dados, de como decorre o ambiente em que os dados são recolhidos, etc.

* ***Estratégia de Eliminação/Remoção*** - elimino as observações todas em que faltam dados em determinada coluna.
* ***Estratégia de Imputação*** - podemos usar a média/moda (um valor à nossa escolha, que pode ser calculado com base na dataframe toda ou em parcelas de dados desta), para atribuir aos campos das nossas observações sem valores (nulos ou vazios).


In [18]:
df_IMDB.isnull()

RANK  TITLE  GENRE  DESCRIPTION  DIRECTOR  ACTORS   YEAR  RUNTIME  \
0    False  False  False        False     False   False  False    False   
1    False  False  False        False     False   False  False    False   
2    False  False  False        False     False   False  False    False   
3    False  False  False        False     False   False  False    False   
4    False  False  False        False     False   False  False    False   
..     ...    ...    ...          ...       ...     ...    ...      ...   
995  False  False  False        False     False   False  False    False   
996  False  False  False        False     False   False  False    False   
997  False  False  False        False     False   False  False    False   
998  False  False  False        False     False   False  False    False   
999  False  False  False        False     False   False  False    False   

     RATING  VOTES  REVENUE  METASCORE  
0     False  False    False      False  
1     False  False    False      False  
2     False  False    False      False  
3     False  False    False      False  
4     False  False    False      False  
..      ...    ...      ...        ...  
995   False  False     True      False  
996   False  False    False      False  
997   False  False    False      False  
998   False  False     True      False  
999   False  False    False      False  

[1000 rows x 12 columns]

In [19]:
df_IMDB.isnull().sum()

RANK             0
TITLE            0
GENRE            0
DESCRIPTION      0
DIRECTOR         0
ACTORS           0
YEAR             0
RUNTIME          0
RATING           0
VOTES            0
REVENUE        128
METASCORE       64
dtype: int64

*Conseguimos concluir que temos 128 dados vazios na coluna 'REVENUE' e 64 na coluna 'METASCORE'.*

* **.dropna()** - cria um novo dataframe onde remove todas as observações que tenham um ou mais campos vazios ou nulos em determinada coluna ou colunas.
<ul><li><strong>axis=0</strong> - por defeito assume 0, elimina as observações que contenham campos vazios ou nulos.</li>
<li><strong>axis=1</strong> - outra opção para eliminar colunas que contenham campos vaziosou nulos.</li></ul>

In [20]:
df2_IMDB = df_IMDB.dropna()
df2_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
Index: 838 entries, 0 to 999
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   RANK         838 non-null    int64  
 1   TITLE        838 non-null    object 
 2   GENRE        838 non-null    object 
 3   DESCRIPTION  838 non-null    object 
 4   DIRECTOR     838 non-null    object 
 5   ACTORS       838 non-null    object 
 6   YEAR         838 non-null    int64  
 7   RUNTIME      838 non-null    int64  
 8   RATING       838 non-null    float64
 9   VOTES        838 non-null    int64  
 10  REVENUE      838 non-null    float64
 11  METASCORE    838 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 85.1+ KB


*Como podemos ver a função .dropna() eliminou 162 observações (alguns dos campos vazios nas duas colunas estavam em observações comuns).*

*A nossa cópia df2_IMDB ficou com apenas 838 observações, enquanto a df_IMDB original tem 1000 observações.*

In [21]:
df_IMDB.shape

(1000, 12)

In [22]:
#Elimino a coluna (axis=1) que tenha algum campo com valores vazios ou nulos
df3_IMDB = df_IMDB.dropna(axis=1)
df3_IMDB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   RANK         1000 non-null   int64  
 1   TITLE        1000 non-null   object 
 2   GENRE        1000 non-null   object 
 3   DESCRIPTION  1000 non-null   object 
 4   DIRECTOR     1000 non-null   object 
 5   ACTORS       1000 non-null   object 
 6   YEAR         1000 non-null   int64  
 7   RUNTIME      1000 non-null   int64  
 8   RATING       1000 non-null   float64
 9   VOTES        1000 non-null   int64  
dtypes: float64(1), int64(4), object(5)
memory usage: 78.2+ KB


*Eliminamos as colunas 'REVENUE' e 'METASCORE' que continham campos vazios.*

---

* **.describe()** - função que nos dá informações de estatísticas mais básicas do nosso dataframe.
<ul><li><strong>média</strong> - é uma medida sensível aos valores da amostra, é mais adequada para situações em que os dados são distribuídos mais ou menos de forma uniforme, ou seja, valores sem grandes discrepâncias.</li>
    <li><strong>std (desvio padrão)</strong> - expressa o quanto um conjunto de dados se desvia da média.</li>
    <li><strong>1º Quartil 25%</strong> - 25% dos valores observados são menores ou iguais a este valor.</li>
    <li><strong>2º Quartil 50% (mediana)</strong> - valor central de todos os dados observados (ou média dos dois valores centrais), 50% dos valores observados são menores ou iguais a este valor.</li>
    <li><strong>3º Quartil 75%</strong> - 75% dos valores observados são menores ou iguais a este valor.</li></ul>

In [23]:
df_IMDB.describe()

RANK         YEAR      RUNTIME       RATING         VOTES  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

          REVENUE   METASCORE  
count  872.000000  936.000000  
mean    82.956376   58.985043  
std    103.253540   17.194757  
min      0.000000   11.000000  
25%     13.270000   47.000000  
50%     47.985000   59.500000  
75%    113.715000   72.000000  
max    936.630000  100.000000

In [24]:
df2_IMDB.describe()

RANK        YEAR     RUNTIME      RATING         VOTES  \
count   838.000000   838.00000  838.000000  838.000000  8.380000e+02   
mean    485.247017  2012.50716  114.638425    6.814320  1.932303e+05   
std     286.572065     3.17236   18.470922    0.877754  1.930990e+05   
min       1.000000  2006.00000   66.000000    1.900000  1.780000e+02   
25%     238.250000  2010.00000  101.000000    6.300000  6.127650e+04   
50%     475.500000  2013.00000  112.000000    6.900000  1.368795e+05   
75%     729.750000  2015.00000  124.000000    7.500000  2.710830e+05   
max    1000.000000  2016.00000  187.000000    9.000000  1.791916e+06   

          REVENUE   METASCORE  
count  838.000000  838.000000  
mean    84.564558   59.575179  
std    104.520227   16.952416  
min      0.000000   11.000000  
25%     13.967500   47.000000  
50%     48.150000   60.000000  
75%    116.800000   72.000000  
max    936.630000  100.000000

*Podemos identificar variações bastante significativas na média, desvio padrão, etc. depois de eliminar as observações com campos vazios.*

---

* **.mode()** - representa o valor mais frequente de um conjunto de dados, sendo assim, para defini-la basta observar a frequência com que os valores aparecem.

In [25]:
#Podemos usar estas funções relativamente a uma coluna específica
df_IMDB['RATING'].mode()

0    7.1
Name: RATING, dtype: float64

---

### Colunas

In [26]:
df_IMDB['RATING']

0      8.1
1      7.0
2      7.3
3      7.2
4      6.2
      ... 
995    6.2
996    5.5
997    6.2
998    5.6
999    5.3
Name: RATING, Length: 1000, dtype: float64

In [27]:
#Podemos usar estas funções relativamente a uma coluna específica
df_IMDB['RATING'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1000 entries, 0 to 999
Series name: RATING
Non-Null Count  Dtype  
--------------  -----  
1000 non-null   float64
dtypes: float64(1)
memory usage: 7.9 KB


In [28]:
#Podemos usar estas funções relativamente a uma coluna específica
df_IMDB['RATING'].describe()

count    1000.000000
mean        6.723200
std         0.945429
min         1.900000
25%         6.200000
50%         6.800000
75%         7.400000
max         9.000000
Name: RATING, dtype: float64

In [29]:
df_IMDB['REVENUE']

0      333.13
1      126.46
2      138.12
3      270.32
4      325.02
        ...  
995       NaN
996     17.54
997     58.01
998       NaN
999     19.64
Name: REVENUE, Length: 1000, dtype: float64

In [30]:
df_IMDB['REVENUE'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1000 entries, 0 to 999
Series name: REVENUE
Non-Null Count  Dtype  
--------------  -----  
872 non-null    float64
dtypes: float64(1)
memory usage: 7.9 KB


In [31]:
df_IMDB['REVENUE'].describe()

count    872.000000
mean      82.956376
std      103.253540
min        0.000000
25%       13.270000
50%       47.985000
75%      113.715000
max      936.630000
Name: REVENUE, dtype: float64

*Normalmente o género do filme está bastante relacionado com a receita que este filme obterá.*

*Poderia fazer sentido, optarmos pela média por género de forma a termos um valor com menor erro para incluirmos nos nossos valores nulos ou vazios.*

In [32]:
df_IMDB['GENRE'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: GENRE, dtype: object

In [33]:
df_IMDB['GENRE'].value_counts()

GENRE
Action,Adventure,Sci-Fi    50
Drama                      48
Comedy,Drama,Romance       35
Comedy                     32
Drama,Romance              31
                           ..
Action,Comedy,Family        1
Action,Crime,Fantasy        1
Comedy,Mystery              1
Adventure,Comedy,Horror     1
Comedy,Family,Fantasy       1
Name: count, Length: 207, dtype: int64

In [34]:
#Número de observações com o "sub-género" 'Action,Adventure,Sci-Fi'
df_IMDB['GENRE'].value_counts()['Action,Adventure,Sci-Fi']

50

In [35]:
#Obtermos a observação/ões que tem o genre igual a 'Action,Comedy,Family'
df_IMDB[df_IMDB['GENRE'] == 'Action,Comedy,Family']

RANK   TITLE                 GENRE  \
386   387  Pixels  Action,Comedy,Family   

                                           DESCRIPTION        DIRECTOR  \
386  When aliens misinterpret video feeds of classi...  Chris Columbus   

                                                ACTORS  YEAR  RUNTIME  RATING  \
386  Adam Sandler, Kevin James, Michelle Monaghan,P...  2015      105     5.6   

      VOTES  REVENUE  METASCORE  
386  101092    78.75       27.0

In [36]:
#Observação que tem um 'revenue' igual a zero
df_IMDB[df_IMDB['REVENUE'] == 0]

RANK             TITLE                 GENRE  \
231   232  A Kind of Murder  Crime,Drama,Thriller   

                                           DESCRIPTION      DIRECTOR  \
231  In 1960s New York, Walter Stackhouse is a succ...  Andy Goddard   

                                                ACTORS  YEAR  RUNTIME  RATING  \
231  Patrick Wilson, Jessica Biel, Haley Bennett, V...  2016       95     5.2   

     VOTES  REVENUE  METASCORE  
231   3305      0.0       50.0

**NOTA:**

Neste momento, temos que analisar se este número '0' é um valor real ou se se trata de um erro!!

Neste tipo de configurações temos que questionar e, eventualmente, voltar atrás no processo de recolha/aquisição de dados, caso se tenha verificado em mais situações.

---

### Estratégia de Imputação

In [37]:
#Criamos uma variável que aponta para a coluna revenue da df original
df_revenue = df_IMDB['REVENUE']

In [38]:
revenue_mean = df_revenue.mean()
revenue_mean

82.95637614678898

In [39]:
#Voltamos a verificar onde estão os valores vazios e quantos são
df_IMDB.isna().sum()

RANK             0
TITLE            0
GENRE            0
DESCRIPTION      0
DIRECTOR         0
ACTORS           0
YEAR             0
RUNTIME          0
RATING           0
VOTES            0
REVENUE        128
METASCORE       64
dtype: int64

* **.fillna()** - substituimos os valores vazios por um novo valor

In [40]:
#Substituimos os valores vazios com a média dos valores observados na coluna revenue
#Como fazemos inplace=True vamos alterar a df original
#Se colocarmos inplace=False, criaria uma cópia da coluna com as alterações na variável revenue
df_revenue.fillna(revenue_mean, inplace=True)

df_IMDB.isna().sum()

RANK            0
TITLE           0
GENRE           0
DESCRIPTION     0
DIRECTOR        0
ACTORS          0
YEAR            0
RUNTIME         0
RATING          0
VOTES           0
REVENUE         0
METASCORE      64
dtype: int64

---

### Correlações entre variáveis
* **.corr()** - obter uma nova dataframe com as correlações entre as variáveis da nossa dataframe inicial. Pertende saber se há comportamentos semelhantes entre as variáveis. Quanto mais próximo de 1 ou -1 o coeficente de correlação é, mais forte é a sua correlação.
<ul><li><strong>numeric_only=False</strong> - por defeito assume False, apenas inclui float, int ou booleano.</li>
<li><strong>numeric_only=True</strong> - para obtermos essa correlação entre variáveis de todos os tipos de dados, inclusive strings.</li>

In [47]:
df_IMDB.corr(numeric_only=True)

RANK      YEAR   RUNTIME    RATING     VOTES   REVENUE  \
RANK       1.000000 -0.261605 -0.221739 -0.219555 -0.283876 -0.252996   
YEAR      -0.261605  1.000000 -0.164900 -0.211219 -0.411904 -0.117562   
RUNTIME   -0.221739 -0.164900  1.000000  0.392214  0.407062  0.247834   
RATING    -0.219555 -0.211219  0.392214  1.000000  0.511537  0.189527   
VOTES     -0.283876 -0.411904  0.407062  0.511537  1.000000  0.607941   
REVENUE   -0.252996 -0.117562  0.247834  0.189527  0.607941  1.000000   
METASCORE -0.191869 -0.079305  0.211978  0.631897  0.325684  0.133328   

           METASCORE  
RANK       -0.191869  
YEAR       -0.079305  
RUNTIME     0.211978  
RATING      0.631897  
VOTES       0.325684  
REVENUE     0.133328  
METASCORE   1.000000

---

### Construir um sub-dataframe a partir do dataframe original

In [48]:
df_IMDB[['RANK', 'VOTES']]

RANK   VOTES
0       1  757074
1       2  485820
2       3  157606
3       4   60545
4       5  393727
..    ...     ...
995   996   27585
996   997   73152
997   998   70699
998   999    4881
999  1000   12435

[1000 rows x 2 columns]

In [49]:
#Isolar apenas uma coluna ou série
df_IMDB['RANK']

0         1
1         2
2         3
3         4
4         5
       ... 
995     996
996     997
997     998
998     999
999    1000
Name: RANK, Length: 1000, dtype: int64

---

* **.set_index()** - conseguimos defenir o nosso index como uma coluna específica.

In [61]:
df_IMDB_title = df_IMDB.set_index('TITLE')
df_IMDB_title

RANK                     GENRE  \
TITLE                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               DESCRIPTION  \
TITLE                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     DIRECTOR  \
TITLE                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    ACTORS  \
TITLE                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         YEAR  RUNTIME  RATING   VOTES     REVENUE  METASCORE  
TITLE                                                                          
Guardians of the Galaxy  2014      121     8.1  757074  333.130000       76.0  
Prometheus               2012      124     7.0  485820  126.460000       65.0  
Split                    2016      117     7.3  157606  138.120000       62.0  
Sing                     2016      108     7.2   60545  270.320000       59.0  
Suicide Squad            2016      123     6.2  393727  325.020000       40.0  
...                       ...

* **.loc[*nome_da_observação*]** - utilizamos o nosso índice para fazermos uma seleção.

In [62]:
df_IMDB_title.loc['Split']

RANK                                                           3
GENRE                                            Horror,Thriller
DESCRIPTION    Three girls are kidnapped by a man with a diag...
DIRECTOR                                      M. Night Shyamalan
ACTORS         James McAvoy, Anya Taylor-Joy, Haley Lu Richar...
YEAR                                                        2016
RUNTIME                                                      117
RATING                                                       7.3
VOTES                                                     157606
REVENUE                                                   138.12
METASCORE                                                   62.0
Name: Split, dtype: object

* **.iloc[*numero_observação*]** - utilizamos o índice natural para fazermos uma seleção.

In [66]:
#Selecionamos a mesma observação mas através de atributos diferentes
df_IMDB_title.iloc[2]

RANK                                                           3
GENRE                                            Horror,Thriller
DESCRIPTION    Three girls are kidnapped by a man with a diag...
DIRECTOR                                      M. Night Shyamalan
ACTORS         James McAvoy, Anya Taylor-Joy, Haley Lu Richar...
YEAR                                                        2016
RUNTIME                                                      117
RATING                                                       7.3
VOTES                                                     157606
REVENUE                                                   138.12
METASCORE                                                   62.0
Name: Split, dtype: object